File: java/UrlCache.java
import java.util.ArrayList;
import java.util.TreeSet;

public class UrlCache {
   
    public UrlCache() {
        gettingFromCache = true;
        unique = 0;
        NUM_VALID_CHARS = 64;
        searchStringBuffer = new StringBuffer();
        urls = new ArrayList<TreeSet<String>>();
        for (int i = 0; i < NUM_VALID_CHARS; ++i) {
            urls.add(new TreeSet<String>());
        }
    }

    public void addToCache(final String url) throws Exception {
        this.storeUrl(url);
    }

    public void initGetFromCache() {
        this.reset();
    }

    public final String getFromCache(char c) throws Exception {
        this.searchStringBuffer.append(c);

        String searchString = searchStringBuffer.toString();
        if (this.gettingFromCache) {
            String urlMatch = this.getUrlMatch(searchString);
            if (urlMatch.isEmpty()) {
                this.gettingFromCache = false;
                return searchString;
            }
            return urlMatch;
        } else {
            return searchString;
        }
    }

    private void reset() {
        this.gettingFromCache = true;
        this.searchStringBuffer.setLength(0);
    }

    private boolean storeUrl(final String urlToCache) throws Exception {
        if (urlToCache.length() == 0) {
            return false;
        }
        int urlsIndex = getVecIndex(urlToCache.charAt(0));
        TreeSet<String> urlSet = this.urls.get(urlsIndex);

        if (urlSet.contains(urlToCache)) {
            System.out.println(urlToCache + " already exists");
            return false;
        }

        urlSet.add(urlToCache);
        ++this.unique;
        return true;
    }

    private String getUrlMatch(final String targetSubstr) throws Exception {
        int urlsIndex = getVecIndex(targetSubstr.charAt(0));
        TreeSet<String> urlSet = this.urls.get(urlsIndex);

        for (String url: urlSet) {
            if (url.length() < targetSubstr.length() ) {
                continue;
            }
            String urlTargetSubstr = url.substring(0, targetSubstr.length());

            if (urlTargetSubstr.compareTo(targetSubstr) == 0) {
                return url;
            }
        }
        return new String();
    }

    private int getVecIndex(char c) throws Exception {

        final int NUM_NUMBERS = 10;
        final int NUM_LETTERS = 26;

        int pos = 0;
        if (c == '-') {
            return pos;
        }

        ++pos;
        if (c == '.') {
            return pos;
        }

        ++pos;
        if (c >= '0' && c <= '9') {
            return pos + c - '0';
        }

        pos += (NUM_NUMBERS - 1);
        if (c >= 'A' && c <= 'Z') {
            return pos + c - 'A';
        }

        pos += (NUM_LETTERS - 1);
        if (c >= 'a' && c <= 'z') {
            return pos + c - 'a';
        }

        if (c == '-') {
            return 0;
        }

        throw new Exception("invalid character");
    }

    private ArrayList<TreeSet<String>> urls;
    private StringBuffer    searchStringBuffer;
    private boolean         gettingFromCache;
    public int              unique;
    private final int       NUM_VALID_CHARS;
};

File: c++/urlCache.cc
#include <iostream>
#include <stdio.h>
#include "urlCache.h"

using std::cout;
using std::endl;

    const unsigned NUM_VALID_CHARS = 64;

    UrlCache::UrlCache() :
        urls(NUM_VALID_CHARS),
        gettingFromCache(true),
        unique(0)
    {}

    void UrlCache::addToCache(const string& url) {
        this->storeUrl(url);
    }

    void UrlCache::initGetFromCache() {
        this->reset();
    }

    const string UrlCache::getFromCache(char c) {
        this->searchString += c;

        if (this->gettingFromCache) {
            string urlMatch;
            if (this->getUrlMatch(searchString, urlMatch)) {
                return urlMatch;
            } else {
                this->gettingFromCache = false;
                return this->searchString;
            }
        } else {
            return this->searchString;
        }
    }

    void UrlCache::reset() {
        this->gettingFromCache = true;
        searchString.clear();
    }

    bool UrlCache::storeUrl(const string& urlToCache) {
        if (urlToCache.length() == 0) {
            return false;
        }
        unsigned urlsIndex = this->getVecIndex(urlToCache[0]);
        set<string, UrlsComp>& urlSet = this->urls[urlsIndex];

        auto findIter = urlSet.find(urlToCache);
        if (findIter != urlSet.end()) {
            cout << urlToCache << " already exists" << endl;
            return false;
        }

        urlSet.insert(urlToCache);
        ++this->unique;
        return true;
    }

    bool UrlCache::getUrlMatch(const string& targetSubstr, string& urlMatch) {
        unsigned urlsIndex = this->getVecIndex(targetSubstr[0]);
        set<string, UrlsComp>& urlSet = this->urls[urlsIndex];

        for (const string& url: urlSet) {
            if (url.length() < targetSubstr.length() ) {
                continue;
            }

            string urlTargetSubstr = url.substr(0, targetSubstr.length());

            if (urlTargetSubstr.compare(targetSubstr) == 0) {
                urlMatch = url;
                return true;
            }
        }
        return false;
    }

    unsigned UrlCache::getVecIndex(char c) {

        const int NUM_NUMBERS = 10;
        const int NUM_LETTERS = 26;

        unsigned pos = 0;
        if (c == '-') {
            return pos;
        }

        ++pos;
        if (c == '.') {
            return pos;
        }

        ++pos;
        if (c >= '0' && c <= '9') {
            return pos + c - '0';
        }

        pos += (NUM_NUMBERS - 1);
        if (c >= 'A' && c <= 'Z') {
            return pos + c - 'A';
        }

        pos += (NUM_LETTERS - 1);
        if (c >= 'a' && c <= 'z') {
            return pos + c - 'a';
        }

        throw string("invalid character");
    }

    unsigned UrlCache::getUniqueUrls() {
        return this->unique;
    }

